slipdev: provide stdio multiplexing over SLIP

miri64 opened the pull request:
### Contribution description
This introduces SLIPMUX like stdio multiplexing to `slipdev`. It can be used in conjunction with `sliptty` from #10477.

### Testing procedure
Test border router application with `USEMODULE += slipdev_stdio` instead of `stdio_ethos` with both `sliptty` and `tunslip`. Both still should work. Also apply the testing procedures from #10477.

Also try to use `tunslip` with just `slipdev` (without `slipdev_stdio`). It should still work as well (just without stdio).

### Issues/PRs references
Requires #10477 for interaction with Linux machine

gebart commented:
Tested gnrc_border_router together with #10477, and it seems to be working well. There is a problem when the Linux host has no global IPv6 address configured which messes up the return routing, see https://github.com/RIOT-OS/RIOT/pull/10477#issuecomment-447761290

miri64 commented:
Rebased to current master.

miri64 commented:
Rebased to current master.

miri64 commented:
Rebased to current master. I think it still needs adaptation for the current `stdio` model.

miri64 commented:
Adapted to current master

benpicco commented:
You can also remove `examples/gcoap/slipdev_params.h` and `tests/slip/slipdev_params.h`

miri64 commented:
These application define the `SLIPDEV` parameters differently (`SLIP_…` instead of `SLIPDEV_…`, so I'd like to keep that for now and tackle this (unrelated) change in a follow-up (if we don't merge #13544 instead)

miri64 commented:
Rebased to current master, fixed the bug discovered in https://github.com/RIOT-OS/RIOT/pull/10477#issuecomment-594195073, and adapted the border router example as promised in https://github.com/RIOT-OS/RIOT/pull/10477#issuecomment-595270792.

miri64 commented:
> You can also remove `examples/gcoap/slipdev_params.h` and `tests/slip/slipdev_params.h`

I've removed the `slipdev_params.h` from the `gcoap` example, but kept the file in the test, to keep the feature of application-side params files in the test-loop.

benpicco commented:
I wasn't so successful this time around.

#### samr21-xpro

```
sudo sh /home/benpicco/dev/RIOT-staging/dist/tools/sliptty/start_network.sh 2001:db8::/64 /dev/ttyACM0 115200
Device "sl0" does not exist.
net.ipv6.conf.all.forwarding = 1
Activated forwarding for all interfaces.
Deactivate with
    sudo sysctl -w net.ipv6.conf.all.forwarding=0
when not desired without this script
net.ipv6.conf.sl0.accept_ra = 2
Starting dispatch. TUN: 3, Stream: 4, In: 0, Out: 1
uhcp_client(): no reply received
uhcp_client(): sending REQ...
got packet from fe80::2360:650e:b9d1:133 port 42209
uhcp: push from fe80::2360:650e:b9d1:133:42209 prefix=2001:db8::/64
gnrc_uhcpc: uhcp_handle_prefix(): add compression context 0 for prefix 2001:db8::80b4:739d:2f34:7a78/64
gnrc_uhcpc: uhcp_handle_prefix(): configured new prefix 2001:db8::80b4:739d:2f34:7a78/64
ifconfig
ifconfig
Iface  7 
          MTU:65535  HL:64  RTR  
          RTR_ADV  
          Link type: wired
          inet6 addr: fe80::2  scope: link  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff00:2
          
Iface  6  HWaddr: 04:45  Channel: 26  Page: 0  NID: 0x23
          Long HWaddr: 82:B4:73:9D:2F:34:7A:78 
           TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
          AUTOACK  ACK_REQ  CSMA  L2-PDU:102 MTU:1280  HL:64  RTR  
          RTR_ADV  6LO  IPHC  
          Source address length: 8
          Link type: wireless
          inet6 addr: fe80::80b4:739d:2f34:7a78  scope: link  VAL
          inet6 addr: 2001:db8::80b4:739d:2f34:7a78  scope: global  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff34:7a78
          
> ping6 fe80::2360:650e:b9d1:133
ping6 fe80::2360:650e:b9d1:133
uhcp_client(): sending REQ...
got packet from fe80::2360:650e:b9d1:133 port 44752
uhcp: push from fe80::2360:650e:b9d1:133:44752 prefix=2001:db8::/64
gnrc_uhcpc: uhcp_handle_prefix(): got same prefix again
gnrc_uhcpc: uhcp_handle_prefix(): add compression context 0 for prefix 2001:db8::80b4:739d:2f34:7a78/64

--- fe80::2360:650e:b9d1:133 PING statistics ---
3 packets transmitted, 0 packets received, 100% packet loss
```

#### Linux

```
enp0s3: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.0.2.15  netmask 255.255.255.0  broadcast 10.0.2.255
        inet6 fe80::4472:e94d:572e:2fe  prefixlen 64  scopeid 0x20<link>
        ether 08:00:27:8f:7e:04  txqueuelen 1000  (Ethernet)
        RX packets 207212  bytes 244058477 (244.0 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 46997  bytes 5940956 (5.9 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 2999  bytes 256665 (256.6 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 2999  bytes 256665 (256.6 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

sl0: flags=4305<UP,POINTOPOINT,RUNNING,NOARP,MULTICAST>  mtu 1500
        inet6 fe80::1  prefixlen 64  scopeid 0x20<link>
        inet6 fe80::2360:650e:b9d1:133  prefixlen 64  scopeid 0x20<link>
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 500  (UNSPEC)
        RX packets 2  bytes 150 (150.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 6  bytes 446 (446.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

% ping6  2001:db8::80b4:739d:2f34:7a78                                                                                  :(
PING 2001:db8::80b4:739d:2f34:7a78(2001:db8::80b4:739d:2f34:7a78) 56 data bytes
^C
--- 2001:db8::80b4:739d:2f34:7a78 ping statistics ---
5 packets transmitted, 0 received, 100% packet loss, time 79ms

% ping6 fe80::80b4:739d:2f34:7a78%sl0 
PING fe80::80b4:739d:2f34:7a78%sl0(fe80::80b4:739d:2f34:7a78%sl0) 56 data bytes
^C
--- fe80::80b4:739d:2f34:7a78%sl0 ping statistics ---
4 packets transmitted, 0 received, 100% packet loss, time 76ms
```

miri64 commented:
When debugging I had some issues with an old-style params file, mapping SLIP to the wrong UART, still hanging around in the BR's directory. Did you check this isn't the case for you?

miri64 commented:
I did bit test with global addresses though as I only realized now, that UHCP not working might be due to my firewall 😅

miri64 commented:
> `> ping6 fe80::2360:650e:b9d1:133`

The interface identifier is missing there ;-)

miri64 commented:
Did you configure a global address for one of your host's interfaces? This worked for me

```
[mlenders@sarajevo RIOT]<3 sudo ip a a affe::1 dev eno1
[mlenders@sarajevo RIOT]<3 ping 2001:db8::28ab:dc15:5401:6478
PING 2001:db8::28ab:dc15:5401:6478(2001:db8::28ab:dc15:5401:6478) 56 data bytes
64 bytes from 2001:db8::28ab:dc15:5401:6478: icmp_seq=1 ttl=64 time=20.3 ms
64 bytes from 2001:db8::28ab:dc15:5401:6478: icmp_seq=2 ttl=64 time=20.3 ms
64 bytes from 2001:db8::28ab:dc15:5401:6478: icmp_seq=3 ttl=64 time=20.3 ms
64 bytes from 2001:db8::28ab:dc15:5401:6478: icmp_seq=4 ttl=64 time=20.3 ms
64 bytes from 2001:db8::28ab:dc15:5401:6478: icmp_seq=5 ttl=64 time=20.2 ms
64 bytes from 2001:db8::28ab:dc15:5401:6478: icmp_seq=6 ttl=64 time=20.3 ms
```

miri64 commented:
(maybe we should add the same hack we use for `start_network.sh` for SLIP as we have for `ethos`, to add a global address to the loopback interface :sweat_smile:)

benpicco commented:
> When debugging I had some issues with an old-style params file, mapping SLIP to the wrong UART, still hanging around in the BR's directory. Did you check this isn't the case for you?

Uh yea I guess it's late…

```
1 benpicco@rechenknecht2k11 ~ % ping6 2001:db8::5806:59c9:2c36:7d78 -s 184                                                                                                   :(
PING 2001:db8::5806:59c9:2c36:7d78(2001:db8::5806:59c9:2c36:7d78) 184 data bytes
192 bytes from 2001:db8::5806:59c9:2c36:7d78: icmp_seq=1 ttl=64 time=42.3 ms
192 bytes from 2001:db8::5806:59c9:2c36:7d78: icmp_seq=2 ttl=64 time=42.5 ms
192 bytes from 2001:db8::5806:59c9:2c36:7d78: icmp_seq=3 ttl=64 time=42.4 ms
192 bytes from 2001:db8::5806:59c9:2c36:7d78: icmp_seq=4 ttl=64 time=42.5 ms
^C
--- 2001:db8::5806:59c9:2c36:7d78 ping statistics ---
4 packets transmitted, 4 received, 0% packet loss, time 3004ms
rtt min/avg/max/mdev = 42.342/42.457/42.540/0.078 ms
benpicco@rechenknecht2k11 ~ % ping6 2001:db8::5806:59c9:2c36:7d78 -s 211 
PING 2001:db8::5806:59c9:2c36:7d78(2001:db8::5806:59c9:2c36:7d78) 211 data bytes
219 bytes from 2001:db8::5806:59c9:2c36:7d78: icmp_seq=1 ttl=64 time=47.4 ms
219 bytes from 2001:db8::5806:59c9:2c36:7d78: icmp_seq=2 ttl=64 time=47.4 ms
219 bytes from 2001:db8::5806:59c9:2c36:7d78: icmp_seq=3 ttl=64 time=47.5 ms
219 bytes from 2001:db8::5806:59c9:2c36:7d78: icmp_seq=4 ttl=64 time=47.4 ms
219 bytes from 2001:db8::5806:59c9:2c36:7d78: icmp_seq=5 ttl=64 time=47.4 ms
```

benpicco commented on drivers/slipdev/stdio.c:
```diff
@@ -0,0 +1,84 @@
+
+void stdio_init(void)
+{
+    /* intentionally overwritten in for netdev init so we have stdio before
```
overwritten in?

benpicco commented on examples/gnrc_border_router/Makefile.slip.conf:
```diff
@@ -1,6 +1,10 @@
 
-CFLAGS += -DSLIP_UART=$(SLIP_UART)
-CFLAGS += -DSLIP_BAUDRATE=$(SLIP_BAUDRATE)
+CFLAGS += -DSLIPDEV_BAUDRATE=$(SLIP_BAUDRATE)
```
```suggestion
CFLAGS += -DSLIPDEV_PARAM_BAUDRATE=$(SLIP_BAUDRATE)
```

benpicco commented on examples/gcoap/Makefile.slip:
```diff
@@ -29,9 +29,8 @@ ifeq (,$(SLIP_BAUDRATE))
-CFLAGS += -DSLIP_UART=$(SLIP_UART)
-CFLAGS += -DSLIP_BAUDRATE=$(SLIP_BAUDRATE)
+CFLAGS += -DSLIPDEV_UART=$(SLIP_UART)
+CFLAGS += -DSLIPDEV_BAUDRATE=$(SLIP_BAUDRATE)
```
```suggestion
CFLAGS += -DSLIPDEV_PARAM_BAUDRATE=$(SLIP_BAUDRATE)
```

benpicco commented on examples/gcoap/Makefile.slip:
```diff
@@ -29,9 +29,8 @@ ifeq (,$(SLIP_BAUDRATE))
-INCLUDES += -I$(CURDIR)
-CFLAGS += -DSLIP_UART=$(SLIP_UART)
-CFLAGS += -DSLIP_BAUDRATE=$(SLIP_BAUDRATE)
+CFLAGS += -DSLIPDEV_UART=$(SLIP_UART)
```
```suggestion
CFLAGS += -DSLIPDEV_PARAM_UART=$(SLIP_UART)
```

miri64 commented:
The input fd is handled last by the `sliptty` tool. That might be the reason for the priorisation.

miri64 commented:
> The input fd is handled last by the `sliptty` tool. That might be the reason for the priorisation.

Oh wait, did you mean there is loss without any load?

miri64 commented:
Interesting observation (not sure it has anything to do with this): all letters of `ifconfig` start with a `0x6` nibble (the signifyer for an IPv6 packet):

```
In [5]: [hex(ord(a)) for a in "ifconfig"]                                                                                                                                                                                             
Out[5]: ['0x69', '0x66', '0x63', '0x6f', '0x6e', '0x66', '0x69', '0x67']
```

miri64 commented:
Squashed

miri64 commented:
Fixed travis errors and squashed.

miri64 commented:
And fixed compile time errors.

miri64 commented:
> (maybe we should add the same hack we use for `start_network.sh` for SLIP as we have for `ethos`, to add a global address to the loopback interface sweat_smile)

See #13572.
